### ISSUE TITLE
 Sketching-based Matrix Computations

### ISSUE THREAD

**Issue by jomsdev** [scipy/scipy#7122]
**TL;DR**: *I think having an implementation for some randomize matrix algorithms would be great. I would like to discuss if the community finds it useful too.*

*Randomized matrix algorithms* have been a hot topic in research the last years. Recent developments have shown their utility in large-scale machine learning and statistical data analysis applications.

Sketching is a way to compress matrices that preserve essential matrix properties. For some problems, sketches can be used to get faster ways to find high-precision solutions to the original problem. This tool can be used for least-squares and robust regression, eigenvector analysis, non-negative matrix factorization, etc...

The main idea of sketching matrices is not new. One of the most famous concepts behind the efficiency of random projection is the **Johnson-Lindenstrauss** lemma. It is used for random projections, and it has a "crude" [implementation in scikit-learn](http://scikit-learn.org/stable/modules/random_projection.html)

An example of sketching is *Uniform Sampling*. It is quick and easy, but doesn't always yield good sketches; however, there are sophisticated sampling methods that do produce good sketches. I have been working on a [xdata](http://www.darpa.mil/program/xdata) open source project for the last year called [libSkylark](https://xdata-skylark.github.io/libskylark/). The library is suitable for general statistical data analysis and optimization applications, but it is heavily focused on distributed systems.

Although many ML applications would take advantage of the implementation of these algorithms the scope of them goes far beyond. **I believe it would be a good idea integrate some of them in Scipy and I would like to know the opinion of others to check if it would be worth it**.

**Edit:** I joined the Scipy dev mailing list yesterday. If the discussion is here successful I would discuss what/how to implement in-depth on the mailing list.

**Comment by rgommers** [comment]
Hi @jomsdev, interesting. First a question on scope fit: you say you're focused on distributed systems, however distributed algorithms and data structures are out of scope for scipy. Are you proposing a non-distributed version here?

Also, how standalone is your code? I see libSkylark having dependencies on FFTW and MPI for example - both of those are showstoppers.

**Comment by jomsdev** [comment]
Hi @rgommers,

Thank you for the fast response. 

As you say, LibSkylark has "hardcore" dependencies on MPI and FFTW. It is a production-ready implementation, and it extraordinarily scales. In fact, this is one of its strength/weakness. I made some python bindings and they are compatible with  numpy/scipy.sparse but the goal of the library is not this one. These methods are really powerful, and I would like having more people with access to them.

For this reason, what I am proposing here is a non-distributed approach.  Some time ago Matlab implemented [Blendenpik](http://epubs.siam.org/doi/10.1137/090767911), an algorithm based on these techniques that outperforms LAPACK by significant factors and scales significantly better than any QR-based solver.

Scipy is one of the most successful libraries in the Scientific Python community and makes sense for me having these methods in it.

**Comment by jomsdev** [comment]
I would like to propose a **clear roadmap** of what/how would be implemented. To my eyes, Randomized Numerical Linear Algebra is big/useful enough to has its submodule. However, this is something that the community has to decide.

Meanwhile, to demonstrate that these methods can be useful I would implement them under `scipy. linalg`,  where the scipy linear algebra functions live.

The idea is:

- Implement [CWT](http://delivery.acm.org/10.1145/2490000/2488620/p81-clarkson.pdf?ip=198.4.83.52&id=2488620&acc=ACTIVE%20SERVICE&key=90259622823F2C32%2E66F8B2C43167A3DD%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&CFID=736713469&CFTOKEN=50269727&__acm__=1489010918_f1abc3d5c4650844f4503dc1a27484ad). **CWT** lets us create a matrix *S* such as *SA* is shorter than *A*.

- Compute *QR = SA* using scipy.

- From here, there are two potential features:

    - Implement [**Blendenpik**](http://epubs.siam.org/doi/10.1137/090767911) for solving Linear Square problems faster without losing any accuracy.

    - Compute the squared row norms to obtain **leverage scores**.

I can implement both but I will start with **Blendenpik**, which is far more common. 

I am thinking about doing it for sparse matrices first because **CWT** can take advantage of it. However, I wonder if the QR method on scipy takes advantage of sparsity too.

I have already set up the development environment so I could start as soon as we decide what to do. As I said I am open to discussion so anyone who could be interested is welcome. 


Thank you,

Jordi.

**Comment by stefanv** [comment]
@jomsdev I can't follow the link to CWT; do you have another?

**Comment by jomsdev** [comment]
@stefanv fixed.

**Comment by rgommers** [comment]
> However, I wonder if the QR method on scipy takes advantage of sparsity too.

No, ``linalg.qr`` doesn't know about sparse matrices.

> To my eyes, Randomized Numerical Linear Algebra is big/useful enough to has its submodule.

I'm not an expert, but it's not obvious from the list above. If it's CWT, Blendenpik and leverage scores, there's no need for a separate submodule right?

These algorithms do look quite promising, and they've been cited enough and been around long enough that it looks to me like they're fine to add to SciPy.

**Comment by jomsdev** [comment]
>  If it's CWT, Blendenpik and leverage scores, there's no need for a separate submodule right?

I took the combination of CWT/Blendenpik because:
- CWT takes advantage of sparsity (can be performed in O(nnz(A))
- CWT can be used as a first step for many algorithms that look like "sketch and solve".
- Blendenpik uses sketching as a preconditioner, but it yields the *real* vector **x**  which solves linear square regression.

However, there are many randomize algorithms for **low-rank approximation, matrix decomposition, eigenvalues, etc**.. that does not yield the "optimal" solution but an approximation. In many cases there is a parameter than can be tuned to achieve the accuracy desired taking into account the trade-off between speed/precision.

If after Blendenpik we see some traction from the community to these methods, I am glad to implement them.

I mentioned the module to be sure that people don't get it wrong (*Why if I execute the same multiple times it gives me a different result?* Or, *Why it does not give the same result than using X algorithm?*)

Anyway, the *random linalg module* (`rlinalg`??) is something that we should not worry about for now (and maybe never!)

**Comment by quattro** [comment]
I think this would be a fantastic feature addition. Us consumers of LA/ML algorithms would greatly benefit from having robust, tested, solutions to sketching problems.

**Comment by jomsdev** [comment]
I already started to implement the [Clarckson-Woodruff Transform in scipy](https://github.com/jomsdev/scipy/tree/CWTransform)

I created a new file under `/scipy/linalg/` called `random_projections.py`. I will implement in it some random projections.

I am now writing some tests (because TDD) but I do not know why I can't execute them. I created a new file [scipy/linalg/tests/test_random_projections.py](https://github.com/jomsdev/scipy/blob/CWTransform/scipy/linalg/tests/test_random_projections.py) following other tests as example but it does not run when I execute the command `python runtests.py -s linalg`

Any idea about why is this happening?

**Comment by perimosocordiae** [comment]
I think your `check_clarkson_woodruff_transform` method needs to start with `test_` to get picked up as a test method. (I.e., use `test_clarkson_woodruff_transform`.)

Also, the current recommendation is to add any new files as private modules (so `_random_projections.py`), to discourage them from being imported directly by users.

**Comment by jomsdev** [comment]
**In order to make the process as accessible to others as I can this message is a copy of the message sent to the official mailing list **

Hello,

I just made a pull request to the official repository with the implementation of the Clarkson Woodruff Transform (Count Min Sketch) for dense matrices.

You can find more details of the pull request [here](https://github.com/scipy/scipy/pull/7809).

I would not like to made it badly (this is my pull request to the project) so any advice is welcome (Main implementation/tests/documentation).

Jordi Montes.

**Comment by rgommers** [comment]
If anyone who's a potential user or is familiar with the algorithm (@quattro, @lesshaste, @mhsouto ?)  would like to review or test the PR (gh-7809), that would be very helpful.

**Comment by quattro** [comment]
I did some basic testing in [REMLSketch](https://github.com/quattro/REMLSketch) and everything ran w/o issue.

I can't speak for future improvements (e.g., sparsity, FLJT, etc), but the current implementation seems to just work.

**Comment by rgommers** [comment]
Thanks @quattro

**Comment by rgommers** [comment]
I'll close this issue, now that we're agreed on it being a good idea to add some of these methods to ``scipy.linalg`` and the first PR being merged. Thanks all!

**Comment by lesshaste** [comment]
It would be good to have a list of PRs here for the interested reader.

**Comment by rgommers** [comment]
@lesshaste sure, we can keep linking to or adding to this issue even when it's closed. Also note that each added function will show up at http://scipy.github.io/devdocs/linalg.html within a day of merging the PR for it.

**Comment by quattro** [comment]
Is there a general consensus/recommendation on what the appropriate sketch size for an n x k matrix? It might be helpful to add some recommendation/suggestion in the docs to guide users not familiar with this. I see mention of an error and suspect its related to those parameters, but again, the connection isn't explicit.
